### PR TITLE
softmmu_template: avoid using shifts and ors in llvm mode

### PR DESCRIPTION
**Pull request by vitalych**
Putting together 2 32-bit values involves an or and a shift,
which produces hard-to-simplify symbolic expressions.
Instead, change the code so that concat expressions are used.

Signed-off-by: Vitaly Chipounov <vitaly@cyberhaven.io>

### CODE REVIEW

File: src/softmmu_template.h [hunk 1]
```diff
@@ -191,7 +191,16 @@ inline DATA_TYPE glue(glue(io_read_chk, SUFFIX), MMUSUFFIX)(ENV_PARAM target_phy
 
 inline DATA_TYPE glue(glue(io_read_chk, SUFFIX), MMUSUFFIX)(ENV_PARAM target_phys_addr_t physaddr, target_ulong addr,
                                                             void *retaddr) {
-    DATA_TYPE res;
+    // Putting together two 32-bit values involves an or and a shift,
+    // which produces hard-to-simplify symbolic expressions.
+    // Instead, we use an union to force casting, which will generate
```
Comment (adrianherrera) [inline src/softmmu_template.h]: "use a union"

File: src/softmmu_template.h [hunk 2]
```diff
@@ -191,7 +191,16 @@ inline DATA_TYPE glue(glue(io_read_chk, SUFFIX), MMUSUFFIX)(ENV_PARAM target_phy
...
+    // simpler concats and extract expression. The union must be volatile
+    // because otherwise the compiler would optimize away this trick.
```
Comment (adrianherrera) [inline src/softmmu_template.h]: "because" is redundant.

The union must be volatile, otherwise the...

**Review by adrianherrera — CHANGES_REQUESTED** [review]
Very minor changes in comments. Code looks fine.

**Comment by vitalych** [comment]
@adrianherrera fixed

**Review by adrianherrera — APPROVED** [review]

**Pull request merged**